### PR TITLE
micro-ROS Benchmarking documentation

### PR DESCRIPTION
**Pull request by Acuadros95**
Signed-off-by: acuadros95 <acuadros1995@gmail.com>

### CODE REVIEW

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+This section will cover micro-ROS memory usage on the most basic entity types. The test on this section have been performed using the provided UDP transport with FreeRTOS + TCP as network stack.
+
+Profiling methodology
+^^^^^^^^^^^^^^^^^^^^^
+
+The memory profile has been performed with the following configuration:
+
+- Reliable entities with a fixed topic size.
+- UDP transport (FreeRTOS + TCP).
```
Comment (pablogs9) [inline docs/rst/microros_documentation/benchmarking/benchmarking.rst:33]: Link to FreeRTOS + TCP page ?

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+.. note::
+
+    .. TODO(acuadros95): Should we explain somewhere the max capacity of a topic regarding MTU, reliability, etc?
+    .. TODO(acuadros95): When is this statement false?
+
+    There are no differences on memory usage between different topic sizes and the reliability kind used, as the topic size plus reliability and/or middleware overhead shall fit in the static buffers pre-allocated by the program at compile-time, defined by the history configuration.
+
+Meanwhile, to measure the different types of memory:
+
+- Static memory: The static memory has been calculated as the difference between the memory occupied by the .bss and .data sections with a non-zero number of entities, and the memory occupied by the same sections when no micro-ROS application is running, that is, the memory occupied by the rest of components of the RTOS and libraries.
```
[the remark this answers is not in the document]
Comment (pablogs9) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst:47]: And data also

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+- USB-CDC: 115200 baudrate.
+- Serial UART: 115200 baudrate.
+- TCP (AWS Secure Sockets) based on `Wi-Fi-Pmod-Expansion-Board <https://www.renesas.com/eu/en/products/microcontrollers-microprocessors/ra-cortex-m-mcus/wi-fi-pmod-expansion-board-80211bgn-24g-wi-fi-pmod-expansion-board>`_.
+
+  - PMOD baudrate: 460800 bauds.
+
+.. figure:: /rst/figures/micro/benchmarking/stream_transports.png
+    :align: center
+
+As expected, USB is the clear winner versus a slower Serial UART and the TCP protocol over a PMOD WiFi module.
```
Comment (pablogs9) [inline docs/rst/microros_documentation/benchmarking/benchmarking.rst:119]: Do not use winner. Maybe something like:

USB shows the higher throughput due to the fact that has the higher bandwidth, followed by TCP and Serial.

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+The tested stream oriented transports and their configuration are:
+
+- USB-CDC: 115200 baudrate.
+- Serial UART: 115200 baudrate.
+- TCP (AWS Secure Sockets) based on `Wi-Fi-Pmod-Expansion-Board <https://www.renesas.com/eu/en/products/microcontrollers-microprocessors/ra-cortex-m-mcus/wi-fi-pmod-expansion-board-80211bgn-24g-wi-fi-pmod-expansion-board>`_.
+
+  - PMOD baudrate: 460800 bauds.
+
+.. figure:: /rst/figures/micro/benchmarking/stream_transports.png
```
Comment (pablogs9) [inline docs/rst/microros_documentation/benchmarking/benchmarking.rst:116]: Just a comment here, what about maintaining the Y axis equally for all three graphs and setting the scale to log? So comparison will be easier. Same for the next graph. 

If is does not look good, ignore this comment.

**Review by pablogs9 — COMMENTED** [review]
In general, apply some bold and italics along the text to emphasize the important names, and parts and to avoid plain paragraphs.

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst (hunk 1 above)
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst]: https://github.com/eProsima/vulcanexus/pull/105/commits/db5c92b76a07451caa8aad2efad2fa99305a7004

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+Profiling methodology
+^^^^^^^^^^^^^^^^^^^^^
+
+The memory profile has been performed with the following configuration:
+
+- Reliable entities with a fixed topic size.
+- UDP transport (FreeRTOS + TCP).
+- Transport MTU: 512 B.
+- XRCE-DDS history: 4 slots.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst]: https://github.com/eProsima/vulcanexus/pull/105/commits/db5c92b76a07451caa8aad2efad2fa99305a7004

[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst]: https://github.com/eProsima/vulcanexus/pull/105/commits/db5c92b76a07451caa8aad2efad2fa99305a7004

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst]: https://github.com/eProsima/vulcanexus/pull/105/commits/db5c92b76a07451caa8aad2efad2fa99305a7004

[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst]: https://github.com/eProsima/vulcanexus/pull/105/commits/db5c92b76a07451caa8aad2efad2fa99305a7004

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+.. note::
+
+    Notice that each of these entities has its own associated topic, concluding that the number of topics used does not impact memory usage.
+
+The total memory (static plus stack plus dynamic) occupied is summarized in the plots below:
+
+.. figure:: /rst/figures/micro/benchmarking/pub_sub_total.png
+    :align: center
+
+From this data, its concluded that a publisher takes a total of ~ 550 B meanwhile a subscriber uses ~ 600 B. There is virtually no difference between these two entities, as the memory pools of micro-ROS RMW are shared among all the entities participating in a given application.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst]: https://github.com/eProsima/vulcanexus/pull/105/commits/db5c92b76a07451caa8aad2efad2fa99305a7004

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+As concluded on the previous section, the memory used is almost identical for a ~ 500 B usage by both entity kinds. Note that it is also virtually identical to the memory used by a publisher or subscriber application.
+
+RMW History
+^^^^^^^^^^^
+
+As explained before, the topic memory comes from the RMW history, which is formed by static memory pools defined on compilation time. 
+For a varying ``RMW_UXRCE_MAX_HISTORY`` between 1 and 10:
+
+.. figure:: /rst/figures/micro/benchmarking/rmw_history.png
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst:95]: https://github.com/eProsima/vulcanexus/pull/105/commits/db5c92b76a07451caa8aad2efad2fa99305a7004

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst (hunk 4 above)
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst:119]: Updated here: https://github.com/eProsima/vulcanexus/pull/105/commits/c721a09b1f5450e9f298d8738b5cb92f11d945b2

Its improves the visual comparison.

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+Publishers and subscribers have been tested varying the ``RMW_UXRCE_MAX_SUBSCRIPTIONS`` and ``DRMW_UXRCE_MAX_PUBLISHERS`` configuration between 1, 5 and 10. The entities are then initialized and used as usual on a ``std_msgs/msg/Int32`` topic.
+
+.. note::
+
+    Notice that each of these entities has its own associated topic, concluding that the number of topics used does not impact memory usage.
+
+The total memory (static plus stack plus dynamic) occupied is summarized in the plots below:
+
+.. figure:: /rst/figures/micro/benchmarking/pub_sub_total.png
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst:66]: https://github.com/eProsima/vulcanexus/pull/105/commits/c721a09b1f5450e9f298d8738b5cb92f11d945b2

The Service/Client graph is not modified, as it includes all types of memory plus the total sum

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+- Transport MTU: 512 B.
+- XRCE-DDS history: 4 slots.
+- RMW History: 4 slots (Except for RMW History section).
+
+.. note::
+
+    .. TODO(acuadros95): Should we explain somewhere the max capacity of a topic regarding MTU, reliability, etc?
+    .. TODO(acuadros95): When is this statement false?
+
+    There are no differences on memory usage between different topic sizes and the reliability kind used, as the topic size plus reliability and/or middleware overhead shall fit in the static buffers pre-allocated by the program at compile-time, defined by the history configuration.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst:42]: https://github.com/eProsima/vulcanexus/pull/105/commits/7bc021dde7b4c84194bdd2a9d62a5bb87f03c5e1

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst (hunk 3 above)
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst]: https://github.com/eProsima/vulcanexus/pull/105/commits/7bc021dde7b4c84194bdd2a9d62a5bb87f03c5e1

File: docs/rst/microros_documentation/benchmarking/benchmarking.rst [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+- Stack memory: The stack consumed during the program execution is taken into account by means of a FreeRTOS specific function involved in the `memory management capabilities offered by this RTOS <https://www.freertos.org/2020/09/micro-ros-on-freertos.html>`__, the `uxTaskGetStackHighWaterMark() <https://www.freertos.org/uxTaskGetStackHighWaterMark.html>`__ function. This function returns the amount of stack that remains unused when the stack consumed by the program is at its greatest value. By subtracting this figure to the total stack available, which is known, one can obtain the stack peak used by the app.
+
+- Dynamic Memory: This is the memory dynamically allocated by the program by calls to ``calloc()`` and ``malloc()`` functions in the C language. To measure it we have hijacked the call to dynamic memory related functions since the ROS 2 stack allows users to feed the program with custom memory allocators.
```
[the remark this answers is not in the document]
Comment (Acuadros95) [reply docs/rst/microros_documentation/benchmarking/benchmarking.rst]: https://github.com/eProsima/vulcanexus/pull/105/commits/7bc021dde7b4c84194bdd2a9d62a5bb87f03c5e1

**Review by pablogs9 — APPROVED** [review]

**Pull request merged**